src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+        if isinstance(rucioClient, Client):
+            rucioClient.delete_replication_rule(rid)
+            rses = [r['rse'] for r in rucioClient.list_rses(rdoc['rse_expression'])]
+        else:
+            rucioClient.deleteRule(rid)
+            rses = [rse for rse in rucioClient.evaluateRSEExpression(rdoc['rse_expression'])]
+
+        # remove the rule id from ruleIds (if any) and remove the RSE name from currentRSEs (if any)
+        if rid in doc['ruleIds']:
+            doc['ruleIds'].remove(rid)
+            modify = True
+            msg = f"remove rid {rid}"
+            report.addEntry('inactive', uuid, msg)
+        for rse in rses:
+            if rse in doc['currentRSEs']:
+                doc['currentRSEs'].remove(rse)
+                modify = True
+                msg = f"remove rse {rse}"
+                report.addEntry('inactive', uuid, msg)
+
+    # make a log record if the DID + Rucio account tuple does not have any existent rules
+    if not rules:
+        msg = f"inactive task {uuid}, container: {pname} for Rucio account {rucioAccount} does not have any existing rules, proceed without update"
+        logger.info(msg)
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info(f"inactive task {uuid}, update {pname}")
+        mgr.updatePileup(doc)
+        msg = f"update pileup {pname}"
+        report.addEntry('inactive', uuid, msg)
+
+
+def activeTask(doc, spec):
+    """
+    Perform single active task over provided MSPileup document
+
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
+    logger = spec.get('logger', getMSLogger(False))
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    marginSpace = spec.get('marginSpace', 0)
+
+    # extract relevant part of our pileup document we'll use in our logic
+    expectedRSEs = set(doc['expectedRSEs'])
+    currentRSEs = set(doc['currentRSEs'])
+    pileupSize = doc['pileupSize']
+    pname = doc['pileupName']
+    inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+    modify = False
+    localCopyOfCurrentRSEs = list(currentRSEs)
+
+    # if expectedRSEs is different than currentRSEs, then further data placement is required
+    # it's possible that data removal is required!
+    if expectedRSEs != currentRSEs:
+        msg = f"active task {uuid}, further data placement required for pileup name: {pname}, with expectedRSEs: {expectedRSEs} and data currently available at: {currentRSEs}"
+        logger.warning(msg)
+
+        # get list of replication rules for our scope, pileup name and account
+        if isinstance(rucioClient, Client):
+            rules = [r for r in rucioClient.list_replication_rules(inputArgs)]
+        else:
+            rules = [r for r in rucioClient.listDataRules(pname)]
```

Comment:
Same comment as before. Missing kwargs and recreating the list already returned.

Reply:
done